File: src/claude_client.py
"""
Claude API Client
Handles communication with Anthropic's Claude API for task execution.
"""

import anthropic
from typing import Dict, Any, Optional, List
import logging

logger = logging.getLogger(__name__)


class ClaudeClient:
    """Client for interacting with Claude API."""

    def __init__(self, api_key: str, model: str = "claude-sonnet-4-20250514"):
        """
        Initialize Claude client.

        Args:
            api_key: Anthropic API key
            model: Claude model to use
        """
        self.client = anthropic.Anthropic(api_key=api_key)
        self.model = model
        self.max_tokens = 4096

    def test_connection(self) -> bool:
        """
        Test connection to Claude API.

        Returns:
            True if connection successful, False otherwise
        """
        try:
            # Send a simple test message
            response = self.client.messages.create(
                model=self.model,
                max_tokens=10,
                messages=[{"role": "user", "content": "Hello"}]
            )
            return response.content is not None
        except Exception as e:
            logger.error(f"Claude API connection test failed: {e}")
            return False

    def execute_task(self, task_description: str, context: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        """
        Execute a task using Claude API.

        Args:
            task_description: Description of the task to execute
            context: Optional additional context for the task

        Returns:
            Dictionary containing execution results with keys:
            - success: bool
            - response: str (Claude's response)
            - usage: dict (token usage)
            - error: str (optional, if failed)
        """
        try:
            # Build the prompt
            prompt = self._build_prompt(task_description, context)

            logger.info(f"Sending task to Claude API (model: {self.model})")

            # Call Claude API
            response = self.client.messages.create(
                model=self.model,
                max_tokens=self.max_tokens,
                messages=[{"role": "user", "content": prompt}]
            )

            # Extract text content
            response_text = ""
            if response.content:
                for block in response.content:
                    if hasattr(block, 'text'):
                        response_text += block.text

            # Calculate costs (approximate)
            usage = {
                'input_tokens': response.usage.input_tokens,
                'output_tokens': response.usage.output_tokens,
                'total_tokens': response.usage.input_tokens + response.usage.output_tokens
            }

            logger.info(f"Task executed successfully. Tokens used: {usage['total_tokens']}")

            return {
                'success': True,
                'response': response_text,
                'usage': usage,
                'model': self.model
            }

        except anthropic.APIError as e:
            logger.error(f"Claude API error: {e}")
            return {
                'success': False,
                'error': str(e),
                'response': ''
            }
        except Exception as e:
            logger.error(f"Unexpected error during task execution: {e}")
            return {
                'success': False,
                'error': str(e),
                'response': ''
            }

    def _build_prompt(self, task_description: str, context: Optional[Dict[str, Any]] = None) -> str:
        """
        Build a prompt for Claude from task description and context.

        Args:
            task_description: The task to execute
            context: Optional additional context

        Returns:
            Formatted prompt string
        """
        prompt_parts = [
            "You are ZephyrOS Executor, an AI assistant that completes coding and development tasks.",
            "",
            "TASK:",
            task_description,
        ]

        if context:
            prompt_parts.extend([
                "",
                "ADDITIONAL CONTEXT:",
            ])
            for key, value in context.items():
                prompt_parts.append(f"{key}: {value}")

        prompt_parts.extend([
            "",
            "Please complete this task and provide detailed output including:",
            "1. Your approach and reasoning",
            "2. Any code or artifacts generated",
            "3. Next steps or recommendations",
        ])

        return "\n".join(prompt_parts)

    def stream_task_execution(self, task_description: str, context: Optional[Dict[str, Any]] = None):
        """
        Execute a task with streaming response (for future real-time UI updates).

        Args:
            task_description: Description of the task to execute
            context: Optional additional context

        Yields:
            Text chunks as they arrive from Claude
        """
        try:
            prompt = self._build_prompt(task_description, context)

            with self.client.messages.stream(
                model=self.model,
                max_tokens=self.max_tokens,
                messages=[{"role": "user", "content": prompt}]
            ) as stream:
                for text in stream.text_stream:
                    yield text

        except Exception as e:
            logger.error(f"Streaming execution error: {e}")
            yield f"Error: {str(e)}"

File: ZephyrOS Executor/ZephyrOS Executor/Core/Services/API/ClaudeClient.swift
//
//  ClaudeClient.swift
//  ZephyrOS Executor
//
//  Client for Anthropic Claude API
//

import Foundation

class ClaudeClient {
    private let apiKey: String
    private let model: String
    private let maxTokens: Int
    private let session: URLSession
    private let apiURL = URL(string: "https://api.anthropic.com/v1/messages")!

    init(apiKey: String, model: String = "claude-sonnet-4-20250514", maxTokens: Int = 4096) {
        self.apiKey = apiKey
        self.model = model
        self.maxTokens = maxTokens

        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 60
        config.timeoutIntervalForResource = 600
        self.session = URLSession(configuration: config)
    }

    // MARK: - API Methods

    func testConnection() async throws -> Bool {
        // Send a minimal test request
        let testRequest = ClaudeRequest(
            model: model,
            maxTokens: 10,
            messages: [ClaudeMessage(role: "user", content: "Hello")]
        )

        do {
            _ = try await sendRequest(testRequest)
            return true
        } catch {
            return false
        }
    }

    func executeTask(description: String, context: [String: Any]? = nil) async throws -> TaskResult {
        let startTime = Date()
        let prompt = buildPrompt(description: description, context: context)

        let request = ClaudeRequest(
            model: model,
            maxTokens: maxTokens,
            messages: [ClaudeMessage(role: "user", content: prompt)]
        )

        let response = try await sendRequest(request)

        // Extract text from response
        let responseText = response.content.compactMap { block in
            if block.type == "text", let text = block.text {
                return text
            }
            return nil
        }.joined()

        let executionTime = Date().timeIntervalSince(startTime)

        return TaskResult(
            response: responseText,
            usage: TokenUsage(
                inputTokens: response.usage.inputTokens,
                outputTokens: response.usage.outputTokens,
                totalTokens: response.usage.inputTokens + response.usage.outputTokens
            ),
            model: model,
            executionTimeSeconds: executionTime
        )
    }

    // MARK: - Helper Methods

    private func sendRequest(_ request: ClaudeRequest) async throws -> ClaudeResponse {
        var urlRequest = URLRequest(url: apiURL)
        urlRequest.httpMethod = "POST"
        urlRequest.setValue(apiKey, forHTTPHeaderField: "x-api-key")
        urlRequest.setValue("2023-06-01", forHTTPHeaderField: "anthropic-version")
        urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let encoder = JSONEncoder()
        encoder.keyEncodingStrategy = .convertToSnakeCase
        urlRequest.httpBody = try encoder.encode(request)

        let (data, response) = try await session.data(for: urlRequest)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw ClaudeError.invalidResponse
        }

        guard httpResponse.statusCode == 200 else {
            throw ClaudeError.httpError(statusCode: httpResponse.statusCode)
        }

        let decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        return try decoder.decode(ClaudeResponse.self, from: data)
    }

    private func buildPrompt(description: String, context: [String: Any]?) -> String {
        var parts = [
            "You are ZephyrOS Executor, an AI assistant that completes coding and development tasks.",
            "",
            "TASK:",
            description
        ]

        if let context = context, !context.isEmpty {
            parts.append("")
            parts.append("ADDITIONAL CONTEXT:")
            for (key, value) in context {
                parts.append("\(key): \(value)")
            }
        }

        parts.append("")
        parts.append("Please complete this task and provide detailed output including:")
        parts.append("1. Your approach and reasoning")
        parts.append("2. Any code or artifacts generated")
        parts.append("3. Next steps or recommendations")

        return parts.joined(separator: "\n")
    }
}

// MARK: - Request/Response Models

struct ClaudeRequest: Encodable {
    let model: String
    let maxTokens: Int
    let messages: [ClaudeMessage]
}

struct ClaudeMessage: Encodable {
    let role: String
    let content: String
}

struct ClaudeResponse: Decodable {
    let id: String
    let type: String
    let role: String
    let content: [ContentBlock]
    let model: String
    let usage: Usage

    struct ContentBlock: Decodable {
        let type: String
        let text: String?

        init(from decoder: Decoder) throws {
            let container = try decoder.container(keyedBy: CodingKeys.self)
            type = try container.decode(String.self, forKey: .type)
            text = try container.decodeIfPresent(String.self, forKey: .text)
        }

        enum CodingKeys: String, CodingKey {
            case type, text
        }
    }

    struct Usage: Decodable {
        let inputTokens: Int
        let outputTokens: Int
    }
}

// MARK: - Errors

enum ClaudeError: LocalizedError {
    case invalidResponse
    case httpError(statusCode: Int)
    case decodingError(Error)

    var errorDescription: String? {
        switch self {
        case .invalidResponse:
            return "Invalid response from Claude API"
        case .httpError(let code):
            return "Claude API error: HTTP \(code)"
        case .decodingError(let error):
            return "Failed to decode response: \(error.localizedDescription)"
        }
    }
}
